src/lib/asn1/asn1_str.cpp

```diff
@@ -58,46 +61,69 @@ ASN1_Tag choose_encoding(const std::string& str,
 
 }
 
+template <typename CharT, class AllocT>
+static std::string ucsX_to_utf8(const std::vector<byte, AllocT> &ucsX)
+   {
+   if (ucsX.size() % sizeof(CharT) != 0)
+      {
+      throw Invalid_Argument("cannot decode UCS string (wrong byte count)");
+      }
+
+   union
+      {
+      const byte  *as_char;
+      const CharT *as_wide_char;
+      };
+
+   as_char = ucsX.data();
+   const size_t wide_char_count = ucsX.size() / sizeof(CharT);
+
+   using converter_t = std::codecvt_utf8<CharT, 0x10ffff, std::consume_header>;
+   std::wstring_convert<converter_t, CharT> convert;
+   return convert.to_bytes(as_wide_char, as_wide_char + wide_char_count);
+   }
```

Comment:
That's more a request for comment. The STL actually brings a way to convert UTF/UCS back and forth. It's deprecated (without replacement) in C++17 though...